# Objetivos de este trabajo
* ¿Qué se quiere conseguir y por qué?

 En esta tarea nuestro objetivo es conectarnos a una base de datos la cual contiene datos sobre libros, editoriales, autores y calificaciones de clientes y reseñas de libros en "POSTGRESQL" con el objetivo de realizar consultas que nos podrian ayudar para obtener información que se utilizará para generar una propuesta de valor para un nuevo producto.

 Esto debido a que con la pandemia ocurrida cambio la rutina diaria de todos y todas. Los habitantes de las ciudades ya no pasaban su tiempo libre fuera, yendo a cafés y centros comerciales; sino que más gente se quedaba en casa, leyendo libros. lo cual atrajo  la atención de las startups (empresas emergentes) que se apresuraron a desarrollar nuevas aplicaciones para los amantes de los libros.

 * ¿A quién le interesa lo que produces?

 Esto sin duda le podria interesar a las nuevas empresas emergentes que apuntan a area de los libros virtuales ya que esta en auge por este evento a nivel mundial que ocurrio.

 * ¿Qué decisiones se tomarán de acuerdo a tu análisis?

Se pueden tomar decisiones mas seguras, por ejemplo en como proceder para generar porpuestas de valor en nuevos productos.

# Conectarse a la base de datos

Aqui accederemos a la base de datos en la cual obtendremos la informacion que necesitamos realizar las consultas deseadas.

In [1]:
# importar librerías
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
engine

Engine(postgresql://practicum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-final-project-db)

# Estudiar las tablas

### "Books"

In [3]:
#Veamos todas las tablas en su primeras 5 filas y asignemosle una variable
query = "SELECT * FROM books;"  # Aqui asignaremos una variable con el nombre de la tabla que deseas consultar
books = pd.io.sql.read_sql(query, con=engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### "Rating"

In [4]:
query = "SELECT * FROM ratings;"  # Reemplaza 'tu_tabla' con el nombre de la tabla que deseas consultar
ratings = pd.io.sql.read_sql(query, con=engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

### "Authors"

In [5]:
query = "SELECT * FROM authors;"  # Reemplaza 'tu_tabla' con el nombre de la tabla que deseas consultar
authors = pd.io.sql.read_sql(query, con=engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

### "Reviwes"

In [6]:
query = "SELECT * FROM reviews;"  # Reemplaza 'tu_tabla' con el nombre de la tabla que deseas consultar
reviews = pd.io.sql.read_sql(query, con=engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### "Publisher"

In [7]:
query = "SELECT * FROM publishers;"  # Reemplaza 'tu_tabla' con el nombre de la tabla que deseas consultar
publishers = pd.io.sql.read_sql(query, con=engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Ya pudimos dar un vistazo previo a las tablas con las cuales trabajaremos. Como podemos ver los nombres de las tablas son bastante descriptivos.

#  Consulta SQL

## 1. Encuentra el número de libros publicados después del 1 de enero de 2000.

In [8]:
#Veamos el numero total de libros en la tabla
print("El numero de libros publicados total es de:", books["publication_date"].count())

El numero de libros publicados total es de: 1000


In [9]:
#Ahora filtraremos solo los libros publicados despues del 1 de entero del 2000 
query = "SELECT * " \
"FROM books " \
"WHERE publication_date > '2000-01-01';"
 
#Ejecutamos la consulta y la almacenamos en la variable n_books_2000              
n_books_2000 = pd.io.sql.read_sql(query, con=engine)
print("Los libros publicados despues del 2000-01-01 son:", n_books_2000["publication_date"].count())

Los libros publicados despues del 2000-01-01 son: 819


De los 1000 datos en total que teniamos, al filtrar el numero de libros publicados despues del 1 de enero del 2000 nos quedamos con 819 libros. Lo cual puede ser un buen indicio ya que muchas de las lecturas mas "populares" fueron lanzadas o "relanzadas" con nuevas editoriales despues del 2000.

## 2. Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro.

In [10]:
#Aqui haremos el query solicitado, usaremos la tabla "ratings"
query = """
SELECT 
DISTINCT(book_id) AS unique_book_id,
COUNT(rating_id) AS cnt_rating,
AVG(rating) AS avg_rating
FROM ratings
GROUP BY unique_book_id
ORDER BY avg_rating DESC;
"""
#Almacenamos la consulta en la variable = consulta2
consulta2 = pd.io.sql.read_sql(query, con=engine)
consulta2.head()


unique_book_id  cnt_rating  avg_rating
0              17           4         5.0
1              20           2         5.0
2              55           2         5.0
3              57           2         5.0
4              62           2         5.0

Ahi tenemos el la lista resultante. Veamos mas a profunidad cuantos libros alcanzaron la calificacion maxima, hay que tener ojo porque mucho de estos libros tienen bajas cantidad de calificaciones por lo cual el que tengan la maxima calificacion en promedio no nos indica tanto.

In [11]:
print("El numero de libros con calificaciones maximas es de:",consulta2["avg_rating"][consulta2["avg_rating"]== 5.0].count())

El numero de libros con calificaciones maximas es de: 44


### 3. Identifica la editorial que ha publicado el mayor número de libros con más de 50 páginas (esto te ayudará a excluir folletos y publicaciones similares de tu análisis).

In [12]:
#Aqui tenemos que unir las tablas de "publishers" y "books" para encontrar el query que nos de la informacin que necesitamos.
query = """
SELECT 
publishers.publisher AS publisher, 
COUNT(books.num_pages) AS book_count
FROM publishers 
INNER JOIN books ON books.publisher_id = publishers.publisher_id 
WHERE books.num_pages > 50 
GROUP BY publishers.publisher 
ORDER BY book_count DESC;
"""
#La almacenamos en la variable "consulta3"
consulta3 = pd.io.sql.read_sql(query, con=engine)
consulta3.head()

publisher  book_count
0             Penguin Books          42
1                   Vintage          31
2  Grand Central Publishing          25
3          Penguin Classics          24
4          Ballantine Books          19

La editorial que publico mas libros con mas de 50 paginas fue "Penguin books". Lo cual puede ser muy interesante para tener en cuenta ya que esta editorial le gusta a la gente, con gran ventaja de las otras editoriales que le siguen. lo cual puede ser rentable para una empresa.

### 4. Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.

In [13]:
#Aqui tenemos que unir 3 tablas "authors", "ratings", "books" para hacer el query que nos de la informacion que solicitamos.
query = """
SELECT 
authors.author AS author,
AVG(ratings.rating) AS avg_rating,
COUNT(ratings.rating_id) AS cnt_rating
FROM ratings
INNER JOIN books ON books.book_id = ratings.book_id
INNER JOIN authors ON authors.author_id = books.author_id
GROUP BY authors.author
HAVING COUNT(ratings.rating_id) > 50
ORDER BY avg_rating DESC;
"""

consulta4 = pd.io.sql.read_sql(query, con=engine)
consulta4.head()

author  avg_rating  cnt_rating
0         J.K. Rowling/Mary GrandPré    4.288462         312
1                    Agatha Christie    4.283019          53
2  Markus Zusak/Cao Xuân Việt Khương    4.264151          53
3                     J.R.R. Tolkien    4.240964         166
4           Roald Dahl/Quentin Blake    4.209677          62

El autor que tiene la mas altas calificaciones (solo libros con arriba de 50 reseñas) es: J.K. Rowling/Mary GrandPré. Esat informacion puede ser muy valiosa ya que este autor le saca una diferencia muy grande en cuanto a cantidad de calificaciones, y aun asi es el con mejor promedio en estas calificaciones, importante para tener en cuenta los libros de este autor si se quiere conseguir mas ventas.

### 5. Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

In [14]:
#Usaremos la informacion de las tablas "reviews" y "ratings"
#En la subconsulta usaremos los alias de "rt" para ratings y "r" para reviews
query = """
SELECT AVG(review_count) AS average_reviews 
FROM (SELECT r.username, COUNT(r.review_id) AS review_count  
FROM ratings AS rt
JOIN reviews AS r ON rt.book_id = r.book_id
WHERE rt.username IN (
SELECT username
FROM ratings
GROUP BY username
HAVING COUNT(book_id) > 50)
GROUP BY r.username) AS user_reviews;
"""

consulta5 = pd.io.sql.read_sql(query, con=engine)
consulta5.head()


average_reviews
0         8.076433

Podemos ver que para los usuarios que calificaron mas de 50 libros con reseñas de texto promediaron un numero 8.076433. Lo cual podemos concluir que el numero de porcentajes que hacen calificaciones a los libros mas importantes es relativamente bajo, lo que nos da para pensar que poca gente se da el tiempo de escribir una reseña y que la mayoria prefiere dar una puntuacion numerica y mas rapida.